ql/src/java/org/apache/hadoop/hive/ql/parse/MergeSemanticAnalyzer.java

```diff
@@ -264,27 +248,16 @@ WHEN NOT MATCHED THEN INSERT VALUES (source.a2, source.b2)
     updateOutputs(targetTable);
   }
 
-  /**
-   * This sets the destination name prefix for update clause.
-   * @param insClauseIdx index of insert clause in the rewritten multi-insert represents the merge update clause.
-   * @param rewrittenCtx the {@link Context} stores the prefixes
-   * @return the number of prefixes set.
-   */
-  protected int addDestNamePrefixOfUpdate(int insClauseIdx, Context rewrittenCtx) {
-    rewrittenCtx.addDestNamePrefix(insClauseIdx, Context.DestClausePrefix.UPDATE);
-    return 1;
-  }
-
   /**
    * If there is no WHEN NOT MATCHED THEN INSERT, we don't outer join.
    */
-  private String chooseJoinType(List<ASTNode> whenClauses) {
+  private boolean hasWhenNotMatchedInsertClause(List<ASTNode> whenClauses) {
```

Comment:
COW would be using FULL or LEFT JOIN, so that should be moved to rewriter

Reply:
Implemented Rewriter instead and added MergeStatement to store the parsed parts of merge. COWMergeRewiter should implement its own join type selection logic

Reply:
👍